drivers/adc/adc_mchp_g1.c

```diff
@@ -0,0 +1,911 @@
+			needed_buffer_size);
+		return -ENOMEM;
+	}
+	return 0;
+}
+
+static int adc_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	const struct adc_mchp_dev_config *const dev_cfg = dev->config;
+	struct adc_mchp_dev_data *dev_data = dev->data;
+	int ret;
+	uint32_t channels, channel_count, index;
+
+	if (sequence->channels == 0) {
+		LOG_ERR("No chennels selected!\n");
+		return -EINVAL;
+	}
+
+	/* Set oversampling */
+	ret = adc_set_oversampling(ADC_REGS, sequence->oversampling);
+	if (ret != 0) {
+		LOG_ERR("Invalid oversampling : %d\n", sequence->oversampling);
+		return ret;
+	}
+
+	/* Set Resolution */
+	ret = adc_set_resolution(ADC_REGS, sequence->resolution, sequence->oversampling);
+	if (ret != 0) {
+		LOG_ERR("Invalid resolution : %d\n", sequence->resolution);
+		return ret;
+	}
+
+	/* Verify all requested channels are initialized and store resolution */
+	channels = sequence->channels;
+	channel_count = 0;
+	while (channels) {
+		/* Iterate through all channels and check if they are initialized */
+		index = find_lsb_set(channels) - 1;
+		if (index >= dev_cfg->num_channels) {
+			LOG_ERR("Invalid channel number : %d", index);
+			return -EINVAL;
+		}
+		/* If the channels is not initialized return invalid */
+		if (dev_data->channel_config[index].initialized == false) {
+			LOG_ERR("Channel is not initialized");
+			return -EINVAL;
+		}
+		channel_count++;
+		channels &= ~BIT(index);
+	}
+
+	/* Check buffer */
+	ret = adc_check_buffer_size(sequence, channel_count);
+	if (ret != 0) {
+		LOG_ERR("Check buffer size invalid\n");
+		return ret;
+	}
+
+	/* Store buffer references for use during sampling */
+	dev_data->buffer = sequence->buffer;
+	dev_data->repeat_buffer = sequence->buffer;
+
+	/* At this point we allow the scheduler to do other things while
+	 * we wait for the conversions to complete. This is provided by the
+	 * adc_context functions. However, the caller of this function is
+	 * blocked until the results are in.
+	 */
+	adc_context_start_read(&dev_data->ctx, sequence);
+
+	/* Wait for all ADC conversions to complete before returning, if it's a synchronous
+	 * call
+	 */
+	ret = adc_context_wait_for_completion(&dev_data->ctx);
+
+	return ret;
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_mchp_dev_data *dev_data = CONTAINER_OF(ctx, struct adc_mchp_dev_data, ctx);
+
+	dev_data->channels = ctx->sequence.channels;
+
+	adc_start_channel(dev_data->dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+	struct adc_mchp_dev_data *data = CONTAINER_OF(ctx, struct adc_mchp_dev_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void adc_mchp_isr(const struct device *dev)
+{
+	struct adc_mchp_dev_data *dev_data = dev->data;
+	uint16_t result;
+
+	adc_interrupt_clear(ADC_REGS);
+	result = adc_get_conversion_result(ADC_REGS);
+	*dev_data->buffer++ = result;
```

Comment:
This is hard to read. Let compiler optimize the code.
`	*dev_data->buffer++ = result;`